moses/feature-selection/man/feature-selection.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" Man page for feature-selection
.\"
.\" Copyright (C) 2012, 2014 Linas Vepstas
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.pc
.TH FEATURE-SELECTION 1 "September 9, 2014" "3.2.8" "OpenCog Learning"
.LO 1
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
feature-selection \- dimensional reduction of datasets
.SH SYNOPSIS
.\" The help & version command line
.B feature-selection
.RB \-h | \--help
.br
.B feature-selection
.RB \--version
.br
.\" The general command line
.B feature-selection
.RB \-i
.IR filename
.RB [ \-a
.IR algo ]
.RB [ \-C
.IR num_features ]
.RB [ \-T
.IR threshold ]
.RB [ \-r
.IR random_seed ]
.RB [ \-u
.IR target_feature ]
.RB [ \-Y
.IR ignore_feature ]
.RB [ \-e
.IR force_feature ]
.RB [ \-j
.IR jobs ]
.RB [ \-F
.IR log_file ]
.RB [ \-L ]
.RB [ \-l
.IR loglevel ]
.RB [ \-o
.IR output_file ]
.RB [ \-D
.IR fraction ]
.RB [ \-E
.IR tolerance ]
.RB [ \-U
.IR num_terms ]
.SH DESCRIPTION
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invoke bold face and italics,
.\" respectively.
\fBfeature-selection\fP is a command-line utility for performing
dimensional reduction of large datasets, so as to prepare them 
for a later regression analysis.  It is intended to be used to filter
data before it is fit with the \fBmoses\fP machine-learning program,
but it can be used in more general settings.  

.PP
.\" ============================================================
.SH EXAMPLES
.TP
.BI feature-selection\ \-i raw_data.csv \ \-asmd\ \-C 10 \ \-o reduced.csv \ \-F fs.log

Use the input file \fIraw_data.csv\fP as the input dataset.  The input
file must be a tab-separated or comma-separated table of values.  The 
dependent feature is assumed to be in the first column.  The \fBsmd\fP
algorithm will be used to select 10 different features that are the most
strongly correlated with the dependent feature.  The output will be
written to the file \fIreduced.csv\fP, consisting of the dependent
feature, in the first column, followed by the 10 most predictive
columns.  A logfile of algorithm progress will be written to
\fIfs.log\fP.

.TP
.BI feature-selection\ \-i raw_data.csv \ \-asmd\ \-C 10 \ \-o reduced.csv \ \-u Q3

As above, but use the column labelled \fIQ3\fP as the target feature.
This assumes that the input dataset contains columns labels as the first
row in the dataset.  The selected column labels are reproduced in the
output file.

.PP
.\" ============================================================
.SH OVERVIEW

In regression analysis, one has, in general, a table of values or
\'features\', and the goal of predicting one feature, the \'target\', 
(located in one column of the table) based on the values of the
other features (in other columns).  It is common that only some
of the features are useful predictors of the target, while the
others are mostly just noise.  Regression analysis becomes difficult
if there are too many input variables, and so regression can be improved
simply by filtering out the useless features. Doing this is generically
called \'feature selection\' or \'dimensional reduction\'.  This
program, \fBfeature-selection\fP, does this, using one of several
different algorithms.

In general, feature selection is done by looking for a correlation
between the target feature, and other features.  This correlation
can be computed in a number of different ways; this program uses
mutual information.  A strong correlation between the featureset
and the target variable suggests that the featureset is the 
appropriate one for performing regression analysis.

It can happen that two (or more) features are strongly correlated with
each-other, and that only one of these is needed as a predictor. Adding
any one of these features to the featureset will improve the
correlation considerably, but adding any of the others will have little
or no effect.  Such sets of features are called \'redundant\'.  In some
cases, it can be useful to avoid redundant features.

This program accepts inputs in a fairly generic whitespace or comma
delimited table format: whitespace may consist of repeated tabs or
blanks; values may be comma-separated but padded with blanks.
The input file may contain comment lines; these lines must start with 
a hash-mark (#), exclamation mark (!) or semi-colon (;) as the first
character.

Boolean values may be indicated with any of the characters 0,f,F and 
1,t,T.  Column values must be either boolean or numeric; at this time,
non-numeric data, including dates or columns containing currency symbols
are not supported.

.PP
.\" ============================================================
.SH OPTIONS
.PP
Options fall into three classes: options for controlling input and
output, options for controlling the algorithm behavior, and general
options.

.SS "General options"
.TP
.B \-h, \-\-help
Print command options and exit.
.TP
.BI \-j\  num \fR,\ \fB\-\-jobs= num
Allocate \fInum\fR threads for feature selection.  Using multiple
threads on multi-core processors will speed the feature selection
search.

.TP
.B -\-version
Print program version, and exit.

.PP
.\" ============================================================
.SS "Input and output control options"
These options control how the input and output is managed by the
program.

.TP
.BI \-F\  filename \fR,\ \fB\-\-log\-file= filename
Write debug log traces to \fIfilename\fR. If not specified, traces
are written to \fBfeature-selection.log\fR.  If the \-L option
is specified, this will be used as the base of the filename.
.TP
.BI \-i\  filename \fR,\ \fB\-\-input\-file= filename
The \fIfilename\fR specifies the input data file. The input table must
be in 'delimiter\-separated value' (DSV) format.  Valid separators
are comma (CSV, or comma-separated values), blanks and tabs
(whitespace). Columns correspond to features; there is one sample per
(non-blank) row. Comment characters are hash, bang and semicolon (#!;)
lines starting with a comment are ignored.
The \fB\-i\fR option may be specified multiple times, to indicate multiple
input files. All files must have the same number of columns.
.TP
.BI \-L\fR,\ \fB\-\-log\-file\-dep\-opt
Write debug log traces to a filename constructed from the
\fIlogfile\fP specified in the \fB\-F\fP command, used as a prefix, 
and the other option flags and their values.  The filename will 
be truncated to a maximum of 255 characters.
.TP
.BI \-l\  loglevel \fR,\ \fB\-\-log\-level= loglevel
Specify the level of detail for debug logging. Possible
values for \fIloglevel\fR are \fBNONE\fR, \fBERROR\fR, \fBWARN\fR,
\fBINFO\fR, \fBDEBUG\fR, and \fBFINE\fR. Case does not matter.
Caution: excessive logging detail can lead to program slowdown.
.TP
.BI \-o\  filename \fR,\ \fB\-\-output\-file= filename
Write results to \fIfilename\fR. If not specified, results are written
to \fBstdout\fR.
.TP
.BI \-u\  label \fR,\ \fB\-\-target\-feature= label
The \fIlabel\fR is used as the target feature to fit.  If none is
specified, then the first column is used.  The very first row of the
input file, if it contains non-numeric, non-boolean values, is
interpreted as column labels, as is the common practice for
CSV/DSV file formats.
.TP
.BI \-Y\  label \fR,\ \fB\-\-ignore\-feature= label
Never select the feature \fIlabel\fR.  This feature will be ignored
during all calculations, and thus is useful if the column contains
an unsupported data type.  This option may be used multiple times.
.TP
.BI \-e\  label \fR,\ \fB\-\-force\-feature= label
Always select the feature \fIlabel\fR.  This option may be used
multiple times.
.PP
.\" ============================================================
.SS "Algorithm control options"
These options provide overall control over the algorithm execution.
The most important of these, for controlling behavior, are the
\fB\-a\fR, \fB\-C\fR and \fB\-T\fR flags.

.TP
.BI \-a\  algorithm \fR,\ \fB\-\-algo= algorithm
Select the algorithm to use for feature selection.
Available algorithms include:
.\" See man gtbl for a description of what can follow a .TS
.\" The x format is a gnu extension, and needs a modern gtbl 
.\" in order for it to work. RHEL4 does not have a modern gtbl :-(
.TS
tab (@);
l lx.
\fBsimple\fR@T{
Simple highest mutual information.

This algorithm computes the (simple, 2-point) mutual information (MI)
between the target variable and each feature, and then returns a list
of those features with the highest MI.  The number of features that are
returned must be specified with the \fB\-C\fP option.

The algorithm is 'simple' in that it does not look for any redundancy
or cross-correlation between the various features.  This makes this
algorithm very fast.
T}

\fBsmd\fR@T{
Stochastic Mutual Dependency.

This algorithm searches for the featureset with the highest mutual
information (MI) with regard to the target variable.  It does so by
adding one feature at a time to the featureset, computing the MI
between the target variable and this featureset,
ranking the result, and keeping only the highest-ranked results.
It can be thought of as a kind-of hill-climbing in the space
of mutual information.  This process is repeated until the desired
number of features is found, or until the MI score stops improving.

The maximum number of desired features must be specified with the
\fB\-C\fP option.  The \fB-T\fP option can be used to specify the
minimum desired improvement in the MI score.  That is, the algorithm
keeps adding features to the feature set until the improvement in the MI
score does not exceed this threshold.  Features are added in random
order, so that if there are redundant features, only one will be 
added, depending on the random seed given with the \fB\-r\fP option.

Two features are considered redundant if they are highly correlated,
so that adding either one of the two may improve MI a lot, but adding
the second will not.  Thus, only one is really needed; using the 
\fB\-T\fP option helps eliminate redundant features.

Note that when the number of desired features is large, this algo can
take a very long time to complete.  Although it can be much more accurate
than \fBinc\fP described below, it can be much much slower.
T}

\fBinc\fR@T{
Incremental, Non-Redundant Mutual Information.

Builds a featureset by incrementally adding features with the highest
mutual information with regard to the target.  Features are accepted
only if the mutual information is above a specified threshold. Features
are rejected if they appear to be redundant with others: that is,
if, by their presence, they fail to change the total mutual information
by more than a minimum amount.

One may specify either the number of features to be selected, or
one may specify a general "pressure" to automatically modulate the
number of features found.  That is, one must specify either the
\fB\-C\fP or the \fB\-T\fP option, as otherwise, all features will
be selected.

This algorithm has a runtime approximately equal to the square of the
number of features selected.
T}

\fBrandom\fR@T{
Random selection.

Choose features randomly.  The number of features to choose must be
indicated with the \fP\-C\fP option.
T}

\fBhc\fR@T{
MOSES Hillclimbing.

Not currently supported.
T}
.TE
.TP
.BI \-C\  num_features \fR,\ \fB\-\-target\-size= num_features
Attempt to select at least \fInum_features\fR out of the input set.

This option is ignored unless the chosen algorithm is \fBsmd\fP or
\fBinc\fP.  When the selected algorithm is \fBinc\fP, then this
over-rides the \fB\-T\fP option setting (effectively, the 
\fB\-T\fP option is ignored).

.TP
.BI \-T\  threshold \fR,\ \fB\-\-threshold= threshold
Apply a floating-point threshold for selecting a feature.
A positive value prevents low-scoring features from being 
selected; a value of zero or less will accept all features.

This option is ignored unless the chosen algorithm is 
\fBsmd\fP or \fBinc\fP.

.TP
.BI \-r\  seed \fR,\ \fB\-\-random\-seed= seed
Use \fIseed\fR as the seed value for the pseudo-random number generator.
The various algorithms use the random number generator in different
ways.  The \fPsmd\fP algorithm explicitly shuffles features, so that 
if the dataset contains multiple redundant features, one will be 
chosen randomly.

.\" ============================================================
.SS "Incremental algorithm options"
These options only apply to the \fB\-ainc\fP algorithm.

.TP
.BI \-D\  fraction \fR,\ \fB\-\-inc\-redundant\-intensity= fraction
Threshold fraction used to reject redundant features. If a feature
contributes less than \fIfraction\fP * \fIthreshold\fP to the total
score, it will be rejected from the final feature set.  That is, if
two features are strongly correlated, one should be considered 
redundant; as to which is de-selected will depend on the random-number
generator, i.e. on the random seed specified with the \fB\-r\fP option.

.TP
.BI \-E\  tolerance \fR,\ \fB\-\-inc\-target\-size\-epsilon= tolerance
To be used only with the \fB\-C\fP option.  The incremental algorithm
is not able to directly select a fixed number of features; rather, it
dynamically adjusts the threshold until the desired number of features
results. This option controls the smallest adjustment made.

.TP
.BI \-U\  num_terms \fR,\ \fB\-\-inc\-interaction\-terms= num_terms
The number of variables used in computing the joint entropy.  Normally,
this algorithm never computes the joint entropy of multiple features;
it only considers the effect of a single feature at a time on the
target (that is, it only computes the mutual information between one
feature and the target).  Specifying a number greater than one will
consider the mutual information between multiple features and the
target.  Note that using this calculation is combinatorially more 
computationally expensive, as all possible choices are considered.
That is, (n choose k) choices are considered, where n==number of
features, and k==number of interaction terms.
.PP
.\" ============================================================
.SH TODO
Document the MOSES-algorithm and the options that it takes: -A -c -f -m
-O -s.  These are not documented because the hill-climbing algo is
currently not supported.
.PP
Document -H scorer  -f initial feat 

.SH SEE ALSO
.br
More information is available at
.B http://wiki.moses.org/w/Feature_selection
.SH AUTHORS
.nh
\fBfeature-selection\fP was written by Nil Geisweiller and modified by
Linas Vepstas
.PP
This manual page is being written by Linas Vepstas.


debian/changelog
opencog (0.1.4-0ubuntu1~gama0) hardy; urgency=low

  * New upstream release.

 -- Gustavo Machado Campagnani Gama <gama@vettalabs.com>  Wed, 01 Oct 2008 01:46:45 -0300

opencog (0.1.3-0ubuntu1~gama0) hardy; urgency=low

  * New upstream release.
  * Splitted binary packages into 'libatomspace', 'opencog-server' and
    'opencog-core'.

 -- Gustavo Machado Campagnani Gama <gama@vettalabs.com>  Tue, 23 Sep 2008 17:51:08 -0300

opencog (0.1.2-0ubuntu1~gama0) hardy; urgency=low

  * New release.

 -- Gustavo Machado Campagnani Gama <gama@vettalabs.com>  Fri, 19 Sep 2008 16:15:19 -0300

opencog (0.1.1-0ubuntu1~gama2) hardy; urgency=low

  * Add amd64 as a supported arch.

 -- Gustavo Machado Campagnani Gama <gama@vettalabs.com>  Fri, 19 Sep 2008 03:43:50 -0300

opencog (0.1.1-0ubuntu1~gama1) hardy; urgency=low

  * Initial Release.

 -- Gustavo Machado Campagnani Gama <gama@vettalabs.com>  Fri, 19 Sep 2008 03:18:48 -0300


moses/feature-selection/main/feature-selection.cc
/** feature-selection.cc ---
 *
 * Copyright (C) 2011 OpenCog Foundation
 *
 * Author: Nil Geisweiller
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License v3 as
 * published by the Free Software Foundation and including the exceptions
 * at http://opencog.org/wiki/Licenses
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program; if not, write to:
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <iostream>
#include <fstream>
#include <memory>
#include <stdio.h>

#include <boost/assign/std/vector.hpp> // for 'operator+=()'
#include <boost/range/algorithm/find.hpp>
#include <boost/range/irange.hpp>

#include <moses/util/Logger.h>

#include <moses/comboreduct/table/table.h>
#include <moses/comboreduct/table/table_io.h>
#include <moses/moses/optimization/hill-climbing.h> // for hc_params

#include "feature-selection.h"
#include "../algo/deme_optimize.h"
#include "../algo/incremental.h"
#include "../algo/random.h"
#include "../algo/simple.h"
#include "../algo/stochastic_max_dependency.h"

namespace moses3 {
    
using namespace combo;
using namespace std;
using namespace boost::assign; // bring 'operator+=()' into scope

void err_empty_features(const string& outfile)
{
    // If the dataset is degenerate, it can happen that feature
    // selection will fail to find any features.  Rather than just
    // crashing as a result, write out an empty file, which at least
    // allows downstream processes to infer what happened (i.e. to tell
    // apart this case from an outright crash.)
    logger().warn() << "No features have been selected.";
    cerr << "No features have been selected." << endl;
    if (!outfile.empty()) {
        ofstream out(outfile.c_str());
        out << "# No features have been selected." << endl;
    }
    exit(0);
}

// Score all individual features of Table
vector<double> score_individual_features(const Table& table,
                                         const feature_selection_parameters& fs_params)
{
    typedef set<arity_t> FS;
    CTable ctable = table.compressed();

    fs_scorer<FS> fs_sc(ctable, fs_params);
    vector<double> res;
    boost::transform(boost::irange(0, table.get_arity()), back_inserter(res),
                     [&](arity_t idx) { FS fs = {idx}; return fs_sc(fs); });
    return res;
}

// log the set of features and its number
void log_selected_features(arity_t old_arity, const Table& ftable,
                           const feature_selection_parameters& fs_params)
{
    // log the number selected features
    logger().info("%d out of %d features have been selected",
                  ftable.get_arity(), old_arity);

    // log set of selected feature set
    const string_seq& labs = ftable.itable.get_labels();
    const vector<double>& sco = score_individual_features(ftable, fs_params);
    for (unsigned i=0; i<labs.size(); i++)
    {
        logger().info() << "log_selected_features(): " << labs[i] << " " << sco[i];
    }
}

void write_results(const Table& selected_table,
                   const feature_selection_parameters& fs_params)
{
    Table table_wff = selected_table;
    if (fs_params.output_file.empty())
        ostreamTable(cout, table_wff);
    else
        saveTable(fs_params.output_file, table_wff);
}

feature_set initial_features(const vector<string>& ilabels,
                             const feature_selection_parameters& fs_params)
{
    vector<string> vif; // valid initial features, used for logging
    feature_set res;

    for (const string& f : fs_params.initial_features) {
        size_t idx = distance(ilabels.begin(), boost::find(ilabels, f));
        if(idx < ilabels.size()) { // feature found
            res.insert(idx);
            // for logging
            vif += f;
        }
        else // feature not found
            logger().warn("No such a feature %s in file %s. It will be ignored as initial feature.", f.c_str(), fs_params.input_file.c_str());
    }
    // Logger
    if(vif.empty())
        logger().info("The search will start with the empty feature set");
    else {
        stringstream ss;
        ss << "The search will start with the following feature set: ";
        ostreamContainer(ss, vif, ",");
        logger().info(ss.str());
    }
    // ~Logger
    return res;
}

feature_set_pop select_feature_sets(const CTable& ctable,
                                    const feature_selection_parameters& fs_params)
{
    if (fs_params.algorithm == moses3::hc) {
        // setting moses optimization parameters
        double pop_size_ratio = 20;
        size_t max_dist = 4;
        score_t min_score_improv = 0.0;
        optim_parameters op_params;
        op_params.opt_algo = moses3::hc;
        op_params.pop_size_ratio = pop_size_ratio;
        op_params.terminate_if_gte = fs_params.hc_max_score;
        op_params.max_dist = max_dist;
        op_params.set_min_score_improv(min_score_improv);
        hc_parameters hc_params;
        hc_params.widen_search = fs_params.hc_widen_search;
        hc_params.fraction_of_nn = fs_params.hc_fraction_of_nn;
        hc_params.single_step = false;
        hc_params.crossover = fs_params.hc_crossover;
        hc_params.crossover_pop_size = fs_params.hc_crossover_pop_size;
        hc_params.crossover_min_neighbors = fs_params.hc_crossover_min_neighbors;
        hc_params.prefix_stat_deme = "FSDemes";
        hill_climbing hc(op_params, hc_params);
        return moses_select_feature_sets(ctable, hc, fs_params);
    } else if (fs_params.algorithm == "inc") {
        return incremental_select_feature_sets(ctable, fs_params);
    } else if (fs_params.algorithm == "smd") {
        return smd_select_feature_sets(ctable, fs_params);
    } else if (fs_params.algorithm == "random") {
        return random_select_feature_sets(ctable, fs_params);
    } else if (fs_params.algorithm == "simple") {
        return simple_select_feature_sets(ctable, fs_params);
    } else {
        cerr << "Fatal Error: Algorithm '" << fs_params.algorithm
             << "' is unknown, please consult the help for the "
             << "list of algorithms." << endl;
        exit(1);
        return feature_set_pop(); // to please Mr compiler
    }
}

feature_set select_features(const CTable& ctable,
                            const feature_selection_parameters& fs_params)
{
    feature_set_pop fs_pop = select_feature_sets(ctable, fs_params);
    OC_ASSERT(!fs_pop.empty(), "There might a bug");
    return fs_pop.begin()->second;
}

feature_set select_features(const Table& table,
                            const feature_selection_parameters& fs_params)
{
    CTable ctable = table.compressed();
    if (fs_params.subsampling_ratio < 1.0) {
        logger().debug() << "Subsample table (size = "
                         << ctable.uncompressed_size() << ")";
        subsampleCTable(fs_params.subsampling_ratio, ctable);
        logger().debug() << "New size = " << ctable.uncompressed_size();
    }
    return select_features(ctable, fs_params);
}

void feature_selection(const Table& table,
                       const feature_selection_parameters& fs_params)
{
    // Select the best features
    feature_set selected_features = select_features(table, fs_params);

    // Add the enforced features
    if (!fs_params.force_features_str.empty()) {
        vector<unsigned> force_idxs =
            get_indices(fs_params.force_features_str, table.itable.get_labels());
        selected_features.insert(force_idxs.begin(), force_idxs.end());
    }

    // Write the features (in log and output)
    if (selected_features.empty())
        err_empty_features(fs_params.output_file);
    else {
        Table filt_table = table.filtered(selected_features);
        log_selected_features(table.get_arity(), filt_table, fs_params);
        write_results(filt_table, fs_params);
    }
}

} // ~namespace moses3
